File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        n = len(prices)
        ans = []
        for i in range (0, n):
            discount = False
            for j in range (i+1, n):
                if prices[i] >= prices[j]:
                    ans.append(prices[i] - prices[j])
                    discount = True
                    break
            if not discount:
                ans.append(prices[i])
        return ans

File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        int n = prices.length;
        Stack<Integer> st = new Stack<>();
        int res[] = new int[n];
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && prices[st.peek()] >= prices[i]) {
                int idx = st.pop();
                res[idx] = prices[idx] - prices[i];
            }
            st.push(i);
            res[i] = prices[i];
        }
        return res;
    }
}